### ISSUE TITLE
Best approach to Active Directory integration?

### ISSUE THREAD

**Issue by obosha** [IdentityServer/IdentityServer4#444]
What is the current recommended approach for using AD as identity store?  Or is this not possible today with IS4?

**Comment by leastprivilege** [comment]
dup of https://github.com/IdentityServer/IdentityServer4/issues/178

**Comment by wojciechrak** [comment]
#178 does not contain any specific information regarding this topic, only "enhancement" label.
Are you saying, that what @obosha is trying to do is currently not possible at all? And you're investigating a way to implement it?

**Comment by leastprivilege** [comment]
It should work - but didn't have the time to try it yet. Feel free to try it yourself and report back with more info.

**Comment by pantonis** [comment]
@obosha, @wojciechrak  Did you manage to integrate IdentityServer with AD?

**Comment by Jenan** [comment]
Please how integrate the active direcory to the identityserver4? Thank you

**Comment by Jenan** [comment]
I have found this answer in SO: https://stackoverflow.com/a/43516747
I think that is good starting point for these things.

**Comment by leastprivilege** [comment]
http://docs.identityserver.io/en/release/topics/windows.html


-------
Dominick Baier

**Comment by djosemartine** [comment]
The last link from @leastprivilege is broken.

**Comment by Bogatinovski** [comment]
So is it possible to use Active Directory with Identity Server?

**Comment by brockallen** [comment]
> So is it possible to use Active Directory with Identity Server?

Yes.

**Comment by Bogatinovski** [comment]
> > So is it possible to use Active Directory with Identity Server?
> 
> Yes.

Please let me clarify my case, What I am trying to do is I want to make my IdentityServer act as an Active Directory and connect it with windows computer as such, so that I can log in on my computer using my IdentityServer account. Can I do this with IdentityServer ? I guess IdentityServer needs to implement the LDAP protocol in order to support this scenario. Is my guess right ?

**Comment by Bogatinovski** [comment]
> If you IS is domain joined, then you can use integrated windows auth. The login process in IS4 is entirely up to your code (via the AccountController in our sample UI), and ASP.NET Core framework can support windows auth. If you're not domain joined, then make your login page do LDAP.

What you have described is how IdentityServer can use existing ActiveDirectory to authenticate users who already have accounts in it.
What I want to do is completely the opposite scenario. I want to have my IS to act as a SSO implemented with AspNetIdentity and SQL server storage. So now let's imagine I have this AspNetIdentity database with users and roles and IdentityServer connected to it. I am using the AccountController for creating users, changing passwords etc. 
Now I want to turn this IdentityServer setup into an ActiveDirectory domain. This means that I will be able to use this IdentityServer AD domain and connect it to my local windows computer  so that when I want to log in into my windows computer I can use my account which was created in IdentityServer through the AccountController on the website. 
In other words I want to know how to build a "mini-Active Directory" by using IdentityServer. How is this feasible?

**Comment by brockallen** [comment]
That sounds you're asking if windows can support users logging into windows with a non-windows account.

**Comment by Bogatinovski** [comment]
Yes you are right. My understanding is that you can configure Windows to use users from any Active Directory Service. In this [wikipedia url](https://en.wikipedia.org/wiki/Active_Directory) AD is defined as: 

> A server running Active Directory Domain Services (AD DS) is called a domain controller. It authenticates and authorizes all users and computers in a Windows domain type network...Active Directory uses Lightweight Directory Access Protocol (LDAP) versions 2 and 3

I guess I am asking if I can turn my IdentityServer into this kind of a active directory domain controller, perhaps by implementing the LDAP protocol.

**Comment by brockallen** [comment]
That's a question for Microsoft.

**Comment by TomCJones** [comment]
Fwiw the internal protocol for Windows auth is Kerberos, NTLM or TLS. There's an example of doing that in Microsoft docs. I would not recommend that you go there. IMHO

thx ..Tom (mobile)

On Jan 22, 2019 6:31 PM, Brock Allen <notifications@github.com> wrote:

That's a question for Microsoft.

**Comment by Bogatinovski** [comment]
Thanks for the protocols clarification. I guess you are right about not wanting to go that way. Instead I will settle with using Azure AD B2C service and use their API to manage the Active Directory right from my IdentityServer. With this in place I can then easily join Windows computers in the Azure AD Domain.
